File: data/projects/financial-contagion/load.R
#' @name load.R
#' @title Load the `financial-contagion` project network (R)
#' @description
#'
#' Reads the two CSVs in this folder and builds a directed, weighted igraph
#' object: directed financial exposures (creditor -> debtor) among ~220 firms
#' across three periods of a crisis (before / during / after). Run it straight
#' (`Rscript load.R`) for a quick summary, or `source()` it and call
#' `load_contagion()` to get the graph in your own script.

# 0. Setup ###################################################################
library(igraph)
library(here)

.dir <- function() here::here("data", "projects", "financial-contagion")

#' Load the node table (one row per firm).
load_nodes <- function() {
  read.csv(file.path(.dir(), "nodes.csv"), stringsAsFactors = FALSE)
}

#' Load the edge table (one row per creditor x debtor x period).
load_edges <- function() {
  read.csv(file.path(.dir(), "edges.csv"), stringsAsFactors = FALSE)
}

#' Build the directed, weighted graph.
#'
#' Edges are weighted by `exposure` (dollars at risk). Because the data is
#' temporal (a `period` column: before/during/after), an exposure between the
#' same pair can appear up to 3 times; igraph keeps them as parallel edges, so
#' filter to one `period` first if you want a single-period graph.
load_contagion <- function(nodes = load_nodes(), edges = load_edges()) {
  g <- graph_from_data_frame(d = edges, directed = TRUE, vertices = nodes)
  igraph::E(g)$weight <- igraph::E(g)$exposure
  g
}

# 1. Demo ####################################################################
if (sys.nframe() == 0) {
  cat("\n\U0001F3E6 financial-contagion (R)\n")
  cat("   Directed creditor -> debtor exposures; before / during / after a crisis.\n\n")

  nodes <- load_nodes()
  edges <- load_edges()
  g <- load_contagion(nodes, edges)

  cat(sprintf("✅ Loaded %d nodes (%d banks, %d hedge funds, %d insurers, %d ccps) and %d edges.\n",
              nrow(nodes), sum(nodes$type == "bank"), sum(nodes$type == "hedge_fund"),
              sum(nodes$type == "insurer"), sum(nodes$type == "ccp"), nrow(edges)))
  for (p in c("before", "during", "after")) {
    cat(sprintf("   period %-7s: %d exposures\n", p, sum(edges$period == p)))
  }
  cat(sprintf("\U0001F517 Directed: %s | total exposure: %s\n",
              is_directed(g), format(round(sum(edges$exposure)), big.mark = ",")))
  cat("\U0001F389 Graph ready. Object `g` is a directed, weighted igraph.\n")
}

File: data/projects/financial-contagion/load.py
"""Load the `financial-contagion` project network (Python).

Reads the two CSVs in this folder and builds a directed, weighted python-igraph
object: directed financial exposures (creditor -> debtor) among ~220 firms across
three periods of a crisis (before / during / after). Run it straight
(``python load.py``) for a quick summary, or import ``load_contagion()`` into
your own script.
"""
from __future__ import annotations

from pathlib import Path
import pandas as pd
import igraph as ig

HERE = Path(__file__).resolve().parent


def load_nodes() -> pd.DataFrame:
    """Node table: one row per firm."""
    return pd.read_csv(HERE / "nodes.csv")


def load_edges() -> pd.DataFrame:
    """Edge table: one row per creditor x debtor x period."""
    return pd.read_csv(HERE / "edges.csv")


def load_contagion(nodes: pd.DataFrame | None = None,
                   edges: pd.DataFrame | None = None) -> ig.Graph:
    """Build the directed, weighted graph (edge weight = ``exposure``).

    The data is temporal (a ``period`` column: before/during/after), so an
    exposure between the same pair can appear up to 3 times as parallel edges.
    Filter to one ``period`` first if you want a single-period graph.
    """
    if nodes is None:
        nodes = load_nodes()
    if edges is None:
        edges = load_edges()
    g = ig.Graph.DataFrame(edges, directed=True, vertices=nodes, use_vids=False)
    g.es["weight"] = g.es["exposure"]
    return g


if __name__ == "__main__":
    print("\n\U0001F3E6 financial-contagion (Python)")
    print("   Directed creditor -> debtor exposures; before / during / after a crisis.\n")

    nodes = load_nodes()
    edges = load_edges()
    g = load_contagion(nodes, edges)

    t = nodes["type"].value_counts()
    print(f"✅ Loaded {len(nodes)} nodes "
          f"({t.get('bank',0)} banks, {t.get('hedge_fund',0)} hedge funds, "
          f"{t.get('insurer',0)} insurers, {t.get('ccp',0)} ccps) "
          f"and {len(edges)} edges.")
    for p in ("before", "during", "after"):
        print(f"   period {p:<7}: {(edges['period']==p).sum()} exposures")
    print(f"\U0001F517 Directed: {g.is_directed()} | total exposure: "
          f"{round(edges['exposure'].sum()):,}")
    print("\U0001F389 Graph ready. Object `g` is a directed, weighted igraph.")
